Don't inherit convenience inits if a designated init is missing.

jrose-apple opened the pull request:
(which can happen if an imported class has un-importable initializers)

Our initializer model guarantees that it's safe to inherit convenience initializers when a subclass has implemented all designated initializers, since each convenience initializer will be implemented by calling one of the designated initializers. If one of the designated initializers *can't* be implemented in Swift, however, then inheriting the convenience initializer would not be safe.

This is potentially a source-breaking change, so the importer will only actually record that it failed to import something in when compiling in Swift 4 mode. I also found an unrelated bug about initializer inheritance, which I filed as [SR-4566](https://bugs.swift.org/browse/SR-4566).

I'm using this as groundwork for the "deserialization recovery" I've been doing. I'd *like* to do something better than dropping a designated initializer just because the declaration it overrides has changed, but this is a safe default, at least.

rdar://problem/31563662

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
@swift-ci Please smoke test Linux

CodaFi commented:
@swift-ci please smoke test linux platform

jrose-apple commented:
Review ping, since I think this one could definitely use a second pair of eyes.

DougGregor commented on include/swift/AST/Decl.h:
```diff
@@ -3257,6 +3259,20 @@ class ClassDecl : public NominalTypeDecl {
+  /// This can occur, for example, if the class is an Objective-C class with
+  /// initializers that cannot be represented in Swift.
+  bool hasMissingDesignatedInitializers() const {
+    (void)getMembers();
```
Can we `lookupDirect` on the name `init` instead, so we only deserialize initializers?

jrose-apple replied:
You mean when we actually do that? Yeah, good point.

DougGregor commented on test/ClangImporter/Inputs/custom-modules/ObjCParseExtras.h:
```diff
@@ -107,6 +107,7 @@ __weak id globalWeakVar;
 - (id)getObjectFromVarArgs:(id)first, ...;
 @end
 
+
```
Spurious whitespace addition.

DougGregor reviewed (approved):
Minor comment aside, LGTM!

jrose-apple commented:
@swift-ci Please smoke test

jrose-apple commented:
@swift-ci Please smoke test

jrose-apple committed:
Don't inherit convenience inits if a designated init is missing.

(which can happen if an imported class has un-importable initializers)

Our initializer model guarantees that it's safe to inherit convenience
initializers when a subclass has implemented all designated
initializers, since each convenience initializer will be implemented
by calling one of the designated initializers. If one of the
designated initializers /can't/ be implemented in Swift, however,
then inheriting the convenience initializer would not be safe.

This is potentially a source-breaking change, so the importer will
only actually record that it failed to import something in when
compiling in Swift 4 mode.

rdar://problem/31563662

jrose-apple commented:
@swift-ci Please smoke test
